File: packages/contracts/src/clarification.ts
import { z } from "zod";

import { surfaceSnapshotSchema } from "./surface";

// Clarification prompt (#36). When the intent engine can't confidently bind the
// referent it asks instead of acting (AD5: clarify below threshold). This is the
// structured ask the dashboard renders — distinct from the free-text `reason` on
// a `clarification_required` ResolvedIntent, which stays for logging.

// Why the engine asked. `low_confidence`: one weak target. `ambiguous`: the top
// candidates are too close to choose between. `no_target`: nothing under the ray.
export const CLARIFICATION_REASONS = ["low_confidence", "ambiguous", "no_target"] as const;
export const clarificationReasonSchema = z.enum(CLARIFICATION_REASONS);
export type ClarificationReason = z.infer<typeof clarificationReasonSchema>;

// One disambiguation choice the user can pick. `confidence` is the CALIBRATED
// score (#100), not raw MediaPipe, so the displayed numbers are meaningful.
export const clarificationOptionSchema = z.object({
  targetId: z.string().min(1),
  label: z.string().min(1),
  surface: surfaceSnapshotSchema.optional(),
  confidence: z.number().min(0).max(1),
});
export type ClarificationOption = z.infer<typeof clarificationOptionSchema>;

// The ask: a reason, a human question, and the options to choose from. `options`
// is empty for `no_target` (nothing to pick — the UI shows a re-point prompt).
export const clarificationRequestSchema = z.object({
  reason: clarificationReasonSchema,
  question: z.string().min(1),
  options: z.array(clarificationOptionSchema),
});
export type ClarificationRequest = z.infer<typeof clarificationRequestSchema>;

File: apps/sidecar/DirectorSidecar/DirectorSidecar/Contracts/Clarification.swift
//
//  Clarification.swift
//  DirectorSidecar
//
//  Port of @handsoff/contracts clarification.ts `clarificationRequestSchema` — the structured
//  disambiguation ask the engine emits when it can't confidently bind the referent (AD5:
//  clarify below threshold). Embedded by a `clarification_required` ResolvedIntent.
//

import Foundation

extension Contracts {
    /// Why the engine asked.
    enum ClarificationReason: String, Codable, Sendable, CaseIterable {
        case lowConfidence = "low_confidence"
        case ambiguous
        case noTarget = "no_target"
    }

    /// One disambiguation choice. `confidence` is the calibrated score (#100).
    struct ClarificationOption: Codable, Sendable, Equatable {
        let targetId: String
        let label: String
        let surface: SurfaceSnapshot?
        let confidence: Double
    }

    /// `clarificationRequestSchema`: a reason, a human question, and the options. `options`
    /// is empty for `no_target` (the UI shows a re-point prompt).
    struct ClarificationRequest: Codable, Sendable, Equatable {
        let reason: ClarificationReason
        let question: String
        let options: [ClarificationOption]
    }
}
